sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Auth/ViewControllers/AuthViewController/LoginTableViewController.swift

```diff
@@ -8,48 +8,201 @@
+            self.setLoginInfoText(loginInfo)
+        }
+        
+        core.networkStatusBlock = { status in
+            if self.needReconnect == true && status != NetworkConnectionStatus.notConnection {
+                self.needReconnect = false
+                self.login()
+            } else {
+                updateLoginInfo?(status)
+            }
+        }
+        updateLoginInfo?(core.networkConnectionStatus())
+    }
+    
+    // MARK: - Disable / Enable inputs
+    private func setInputEnabled(enabled: Bool) {
+            chatRoomNameTextField.isEnabled = enabled
+            userNameTextField.isEnabled = enabled
+    }
+    
+    // MARK: UIControl Actions
+    @IBAction func didPressLoginButton(_ sender: QBLoadingButton) {
+        login()
+    }
 
+    // MARK: - UITableViewDelegate
+    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return UITableView.automaticDimension
+    }
+    
+    @IBAction func editingChanged(_ sender: UITextField) {
+        validate(sender)
+        loginButton.isEnabled = userNameIsValid() && chatRoomIsValid()
     }
     
-    // MARK: - Table view data source
-    override func numberOfSections(in tableView: UITableView) -> Int {
-        // #warning Incomplete implementation, return the number of sections
-        return 0
+    func validate(_ textField: UITextField?) {
+        if textField == userNameTextField && !userNameIsValid() {
+            chatRoomDescritptionLabel.text = ""
+            userNameDescriptionLabel.text = LoginConstant.shouldContainAlphanumeric
+        } else if textField == chatRoomNameTextField && !chatRoomIsValid() {
+            userNameDescriptionLabel.text = ""
+            chatRoomDescritptionLabel.text = LoginConstant.shouldContainAlphanumericWithoutSpace
+        } else {
+            userNameDescriptionLabel.text = ""
+            chatRoomDescritptionLabel.text = userNameDescriptionLabel.text
+        }
+        tableView.beginUpdates()
+        tableView.endUpdates()
+    }
+    
+    func setLoginInfoText(_ text: String?) {
+        if text != loginInfo.text {
+            loginInfo.text = text
+            tableView.beginUpdates()
+            tableView.endUpdates()
+        }
+    }
+    
+    // MARK: - Login
+    func login() {
+        isEditing = false
+        beginConnect()
+        if core.currentUser != nil {
+            core.loginWithCurrentUser()
+        } else {
+            core.signUp(withFullName: userNameTextField.text, roomName: chatRoomNameTextField.text)
+        }
+    }
+    
+    func beginConnect() {
+        setInputEnabled(enabled: false)
+        loginButton.showLoading()
+    }
+    
+    func endConnectError() throws {
+        setInputEnabled(enabled: true)
+        loginButton.hideLoading()
+    }
+    
+    // MARK: - Validation helpers
+    func userNameIsValid() -> Bool {
+        let characterSet = CharacterSet.whitespaces
```

Comment:
Wrong internal variables naming!
let characterSet = CharacterSet.whitespaces
        let trimmedText = userNameTextField.text?.trimmingCharacters(in: characterSet)
        let regularExtension = LoginNameRegularExtention.user
        let predicate = NSPredicate(format: "SELF MATCHES %@", regularExtension)
        let isValid = predicate.evaluate(with: trimmedText)
        return isValid